lib/VCMIDirs.cpp

```diff
@@ -164,178 +97,111 @@ class VCMIDirsWIN32 final : public IVCMIDirs
-	// If this parameter is nullptr, retrieves the path of the executable file of the current process.
-	auto getModulePath = [](HMODULE hModule) -> bfs::path
+	if (!bfs::exists(configPath))
 	{
-		wchar_t exePathW[MAX_PATH];
-		DWORD nSize = GetModuleFileNameW(hModule, exePathW, MAX_PATH);
-		DWORD error = GetLastError();
-		// WARN: Windows XP don't set ERROR_INSUFFICIENT_BUFFER error.
-		if (nSize != 0 && error != ERROR_INSUFFICIENT_BUFFER)
-			return bfs::path(std::wstring(exePathW, nSize));
-		// TODO: Error handling
-		return bfs::path();
-	};
+		dirsConfig = std::nullopt;
+		return;
+	}
 
-	// Moves one directory contents to another directory
-	// Shows user the "moving file dialog" and ask to resolve conflicts.
-	// It takes into account that 'from' path can contain current executable.
-	// If necessary closes program and starts update script.
-	auto advancedMoveDirIfExists = [getModulePath, moveDirIfExists](const bfs::path& from, const bfs::path& to) -> bool
+	std::ifstream in(configPath.string(), std::ios::binary);
+	if (!in)
 	{
-		const bfs::path executablePath = getModulePath(nullptr);
-
-		// VCMI can't determine executable path.
-		// Use standard way to move directory and exit function.
-		if (executablePath.empty())
-			return moveDirIfExists(from, to);
+		dirsConfig = std::nullopt;
+		return;
+	}
 
-		const bfs::path executableName = executablePath.filename();
+	std::string buffer((std::istreambuf_iterator<char>(in)), {});
+	dirsConfig = JsonNode(reinterpret_cast<const std::byte*>(buffer.data()), buffer.size(), configPath.string());
+}
 
-		// Current executabl isn't in 'from' path.
-		// Use standard way to move directory and exit function.
-		if (!bfs::equivalent(executablePath, from / executableName))
-			return moveDirIfExists(from, to);
+bfs::path VCMIDirsWIN32::getPathFromConfigOrDefault(const std::string& key, const std::function<bfs::path()>& fallbackFunc) const
+{
+	loadDirsJsonIfNeeded();
 
-		// Try standard way to move directory.
-		// I don't know how other systems, but Windows 8.1 allow to move running executable.
-		if (moveDirIfExists(from, to))
-			return true;
+	if (!dirsConfig || !dirsConfig->isStruct())
+		return fallbackFunc();
 
-		// Start copying script and exit program.
-		if (StartBatchCopyDataProgram(from, to, executableName))
-			exit(ERROR_SUCCESS);
+	const auto& structMap = dirsConfig->Struct();
+	auto it = structMap.find(key);
+	if (it == structMap.end() || !it->second.isString())
+		return fallbackFunc();
 
-		// Everything failed :C
-		return false;
-	};
+	std::wstring raw = utf8ToWstring(it->second.String());
+	wchar_t expanded[MAX_PATH];
+	if (ExpandEnvironmentStringsW(raw.c_str(), expanded, MAX_PATH))
+		return bfs::path(expanded);
+	else
+		return bfs::path(raw);
+}
 
-	moveDirIfExists(oldUserSavePath(), userSavePath());
-	advancedMoveDirIfExists(oldUserDataPath(), userDataPath());
+void VCMIDirsWIN32::init()
+{
+	// Call base (init dirs)
+	IVCMIDirs::init();
```

Comment:
just remove this Windows implementation entirely

Reply:
Will check if can be done without creating empty directories

Reply:
if you don't override method of superclass, it will be called automatically - that's my point